Add has_connection waiting

schaumb opened the pull request:
https://github.com/streamlit/streamlit/issues/2926

Commit:
Add has_connection waiting

Commit:
Add has_connection waiting

Commit:
Stop from signal too

Commit:
Stop from signal too

Commit:
[feature] performance boost

Commit:
[feature] performance boost

Commit:
[refactor] performance boost

Commit:
[refactor] performance boost

schaumb commented:
We create a nicer solution for this 'busy' loop handling

https://github.com/FloWide/streamlit/compare/5f1806525ecac9f4810379b70f2bf285cf91eeb8...FloWide:fbc4952880be530ebaf12bf69f96952cd8b8ad97

This operates only with events, with kept 0.01 sleep on message chunks writing
This solution can be better

ktosiek committed:
Generate the function's cache key lazily (#2798)

* Generate the function's cache key lazily

This way globals that are only created after the cached function
can actually be included in the hash. Before this change only their
names were included, which could lead to stale cache being used.

* Log function's cache_key after it's actually computed

* Test invalidating st.cache() when a dependency changes

* Make test_invalidating_cache more precise

* Better explaination for generating cache_key late

mesmith027 committed:
spin down process added to docs (#2917)

* spin down process added to docs

* Suggested edits to phrasing

Co-authored-by: Marisa Smith <>

dependabot[bot] committed:
Bump react-dev-utils from 11.0.3 to 11.0.4 in /frontend (#2941)

Bumps [react-dev-utils](https://github.com/facebook/create-react-app/tree/HEAD/packages/react-dev-utils) from 11.0.3 to 11.0.4.
- [Release notes](https://github.com/facebook/create-react-app/releases)
- [Changelog](https://github.com/facebook/create-react-app/blob/master/CHANGELOG-1.x.md)
- [Commits](https://github.com/facebook/create-react-app/commits/HEAD/packages/react-dev-utils)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

randyzwitch committed:
Make favicon smaller

AnOctopus commented:
@schaumb Thank you for submitting this!

Would there be a significant performance difference between the two solutions?
Also, do you have any suggestions for testing the performance of the changes, to validate them and help prevent regressions?

Commit:
Combined feature branch for tooltips + theming (#2959)

* Tooltips: add `help` keyword/field to protobufs and API (#2586)

* add help keyword/field to protobufs and api

* fix pylint stuff

* fix tests

* move help to last kwarg

* Add tooltip to TextInput (#2588)

* add help keyword/field to protobufs and api

* add tooltip to TextInput

* remove unnecessary code

* fix pylint

* fix pylint stuff

* fix tests

* remove unnecessary change

* add e2e test

* remove unnecessary change

* see if this fixes some cypress tests

* make requested changes

* make other requested change

* fix failing test

* Add tooltips to even more widgets (#2594)

* add help keyword/field to protobufs and api

* add tooltip to TextInput

* remove unnecessary code

* fix pylint

* fix pylint stuff

* fix tests

* remove unnecessary change

* add e2e test

* remove unnecessary change

* save work

* get tooltip working with button

* remove unnecessary change

* see if this fixes some cypress tests

* make requested changes

* fix merge

* fix imports

* update st_tooltips.spec.js

* destructure

* Add tooltips to more widgets (#2593)

* add help keyword/field to protobufs and api

* add tooltip to TextInput

* remove unnecessary code

* fix pylint

* fix pylint stuff

* fix tests

* remove unnecessary change

* add e2e test

* remove unnecessary change

* add tooltips for st.checkbox, st.radio, st.number_input

* fix import order

* see if this fixes some cypress tests

* make requested changes

* fix

* fix test

* add tests and fix tooltip positioning

* Add tooltips to visual elements (#2614)

* add tooltips to visual elements

* fix newline

* Add missing help keywords and protos (#2620)

* add missing help keywords and protos

* fix test

* add missing function/proto

* fix newline

* Add tooltips to Markdown elements (#2621)

* add tooltips to visual elements

* fix newline

* add missing help keywords and protos

* add new tests

* add remaining tooltips

* add style change

* move changes to a different branch

* add newline

* fix tooltips for a few inputs (#2631)

* Theme selector (#2663)

* Add selector + themes

* Default theme + cleanup

* Rename main theme to light theme (#2670)

* Add customTheme config options (#2671)

* Reorder imports in config.py to be alphabetical

* Allow capital letters in config sections

We don't currently have any two-word config sections but want to add a
`customTheme` section and need it to be camelCase for consistency with
how individual config options are named.

* Add config options for custom themes

* Have customTheme.font default to "sans serif"

* Convert themes to config and add baseweb dark mode (#2675)

* Convert themes to config and add baseweb dark mode

* PR comments - rename + typo

* Add CustomThemeConfig protos and pass theme config to the client (#2681)

* Add CustomThemeConfig proto message type

* Pull some NewReport msg construction into helper functions

Also wrote a test.

* Read customTheme config and include it in Initialize msg

* Promote custom_theme to be a child of the Initialize proto

* Add config.get_options_by_section

* Use config.get_options_by_section to marshal custom_theme protos

* Fix grammatical error in comment

* Rename get_options_by_section -> get_options_for_section

* Add types to new functions in report_session.py

* Reorder imports in report_session.py

* Set theme (#2687)

* Set theme + reorg theme to use ThemeConfig

* PR feedback

* Add serif font (#2691)

* Create custom theme (#2693)

* Create custom theme

* Tests + fixes from tests

* Fix linting

* Fix light + dark theme colors

* Fix sidebar theme + styling

* Fix report_session_test (#2707)

Apparently a constructor signature changed on `develop`, causing a newly added
test in `report_session_test` to fail.

* save work

* save work

* add final change

* Theming Color Hookups (#2710)

* wip

* Update styling variables

* Remove started global styles

* revert unwanted changes

* Theming global styles  (#2714)

* Set up globalStyles

* copy and start updating theme.scss -> emotion

* copy and convert reboot.scss -> emotion

* convert copied sass -> emotion theme

* Update comment

* add tooltips to rest of widgets

* Apply theme to VegaLite Charts (#2708)

* Apply theme to VegaLite Charts

* Substitute theme.colors.bodyText -> textColor and update snapshots

* save work

* make tests pass

* Have plotly charts apply theme options (#2719)

* Destructure theme in VegaLiteChart.tsx

* Have plotly charts apply theme options

This isn't totally done yet as plotly has two background colors: the
"paper" background color and "plot" background color, and the default
plot background color looks fine in light mode but pretty silly in dark.

We could fix this by changing plotly chart style to match our other plot
types, but for now I just left a TODO and can circle back to this later.

* rename config names (#2740)

* Theming sidebar (#2718)

* Remove sidebar theme + cleanup for sidebar

* Unit tests

* Checkbox

* Fix sidebar gradiant

* Update sidebar snapshot + radio button

* VegaLite bg fix

* Calendar header

* Add snapshot

* PR feedback

* Opps

* Theming widget colors (#2742)

* Create widget variables based on provided colors + updates to widgets

* Expander text + updated snapshot

* Theming - available themes (#2744)

* Add auto as available theme + fix constant adding of custom

* Update local store with custom theme if active

* Make sure auto gets latest auto, not local storage

* Fix test

* Do not save auto theme to storage

* Theming widgets (#2745)

* Colors for dropdowns + fullscreen buttons

* Color in stuff for status widget

* Remove unnecessary white color

* Update tests

* Fix typing and remove unnecessary typing. Thanks ken

* Handle test setting Auto with previous saved value (#2747)

* Set plotly chart background to theme secondaryBg color (#2748)

* Set plotly chart background to theme secondaryBg color

Now that the sidebar background color has been renamed/repurposed to be
a secondaryBg color, it seems appropriate to use it for this. This'll
probably need a product review down the line, but this now looks pretty
reasonable in both light and dark mode compared to before.

* Add snapshot test

* Set RNG seed

* Update snapshot

* Theming OS change (#2755)

* Add event listener for OS theme change

* Doing too much reorg, only update auto theme if auto was last selected

* Pass theme object to custom components (#2749)

* Pass theme object to iframe

Note that we still need to update the custom component template along with the
code living in component-lib to use the theme object.

* Pass theme to component and set css vars in component-lib

* Don't set font-family in component-lib css for now

* Change up visual of theme selector (#2759)

* Set custom theme by default if one exists (#2758)

* Add customTheme.setAsDefault config var and pass to frontend

* Set custom theme by default if flag is true

* Theming dark primary (#2763)

* Change up visual of theme selector

* Change dark mode primary color

* Change default theme to create auto theme, not get system theme (#2768)

* Make theme-related types in component-lib more strict (#2762)

* Fixing monospace up (#2769)

* Change default theme to create auto theme, not get system theme

* Rename mono, fix mono, hardcode some things to not mono

* tests

* handle font = 0

* Bump component-lib version (#2772)

* Theming - Fail gracefully for invalid colors (#2771)

* Fail gracefully for hex values

* fix merging test

* Rename config options (#2775)

* Export theme type from component-lib (#2777)

* merge (#2785)

* Validate customTheme config options (#2776)

* Have customTheme.name default to "Custom Theme"

* Validate customTheme config options

This entails throwing an error when
* The theme name specified is reserved ("Auto", "Dark", or "Light")
* The theme config does not specify all required options

* Appease mypy

* Move theme validation into config.py

This needs to happen since validation has to occur when a user calls
`streamlit run <filepath>` rather than on script run.

* Fix some name issues that came up after rebase

* Address PR comments

* Theming docstring (#2788)

* Pick colors for docstring

* Add E2E

* Snapshot

* Remove custom theme option when server doesn't send one (#2787)

* Remove custom theme option when server doesn't send one

* Don't ignore user preferences on theme remove

* Move comment to right place

* Remove unused imports

* Black or white pre based on bg (#2794)

* merge conflicts (#2797)

* Theming settings (#2799)

* Quick styling

* Change radio to dropdown + styling + rename auto

* Add themes to snapshot

* Update active theme on custom theme selection (#2800)

* Save on settings dialog changes and remove save/cancel buttons (#2802)

* Change theme object passed to components to match options in config.py (#2806)

* Have theme passed to components match options in config.py

* Have component-lib use new names

* Theme Creator UI (#2786)

* Quick and dirty prototype

* Adjust color picker UI + add show value option

* Redesign theme creator + copy to clipboard

* Fix tests + typo

* Auto update

* toComponentTheme -> toThemeInput

* Add font

* Tests

* PR feedback

* Add developer mode (#2835)

* Fix font parsing + typo in config copy (#2842)

* Uppercase color picker display (#2850)

* Theming contrast (#2773)

* Add func to ensure contrasting colors

* Color updates

* Color adjustments

* Update snapshots

* Update active theme when editing custom theme (#2845)

* Change to provider + selectbox didupdate

* PR comment

* Promote Config and CustomTheme protos to be children of NewReport (#2843)

These messages were originally children of `NewReport.Initialize`, but as
we're making the change to have (some) config options auto-reload on file
save, this shouldn't be the case anymore given the initialize message by
definition only contains things that never change.

* More font conversions (#2852)

* Fix font conversions

* PR comments

* Check for any merge weirdness merging develop into theming (#2854)

* `watch_file` utility function (#2851)

Adds two file-watching utility functions:

- `get_file_watcher_class()` (returns the correct class to use for watching files, based on the user's config. This has been hoisted out of `local_sources_watcher`).
- `watch_file()` (watches a single file using the correct fileWatcher)

This de-couples file-watching from `LocalSourcesWatcher` so that we can easily use it outside of that context. (Secrets.toml and config.toml reloading will use this code.)

* st.beta_secrets (#2757)

Adds the `st.secrets` module to streamlit (as `st.beta_secrets`).

- `./.streamlit/secrets.toml` is loaded on startup if it exists, and is accessible via `st.beta_secrets`
- If secrets.toml is modified, it will be automatically reloaded
- Top-level key-value pairs in secrets.toml are also added to `os.environ`

* hide empty columns on mobile (#2756)

Co-authored-by: Tim Conkling <tconkling@gmail.com>
Co-authored-by: bh-streamlit <76131205+bh-streamlit@users.noreply.github.com>

* Theming creator v2 (#2853)

* Update copy + edit active theme

* Change on copied

* get rid of props

* Theming: Always pick the user choice upon reload (#2857)

* Update copy + edit active theme

* Change on copied

* get rid of props

* Always take users preference

* Fix tests

* Cleanup

* Theming optional configs (#2859)

* make secondary color optional

* Check theme completeness

* Theming creator copy (#2863)

* make secondary color optional

* Check theme completeness

* Copy changes

* If custom theme active, update in case it changed (#2864)

* Make config.{get_option, set_option, get_options_for_section, parse_config_file} thread-safe (#2860)

* Move variables/consts closer to where they're used in config.py

* Make config.{get_option, set_option, get_options_for_section, parse_config_file} thread-safe

This needs to be done since it'll soon be possible for config files to
be reloaded on file change, so we'll need to ensure that threads trying
to access/change config options don't try to do so when config.toml
files are being re-parsed.

* Double-check if work is needed after grabbing lock

* Rework config file parsing to unset removed options (#2866)

* Rework config file parsing to unset removed options

`parse_config_file` (now named `get_config_options`) was previously
written in a way where naively rerunning the function would fail to
remove config options that were deleted from a file on initial load.

This commit restructures things to create a `_config_options_template`
when options are initially defined, then copy and update the template
when loading config options from files / cli flags.

* Add a missing unit test for get_config_options with flags

* Add assert

* Rerun scripts on config file change (#2869)

* Rerun scripts on config file change

There's still a bit of polish to do here. Most notably, we're currently just
naively reloading the script whenever the config gets re-parsed, which is fine,
but we want to additionally give the user a heads up when they change config
options that require a full server restart.

* Remove redundant call to localStorage.removeItem

* Use secondary color more (#2882)

* Use secondary color more

* Snapshot update

* Relax theme validation to just print warnings instead of raising (#2878)

* Relax theme validation to just print warnings instead of raising

Raising exceptions here now seems like a bad user experience since it's
very likely that the user will get annoyed by things exploding when
they're live-editing config files to see changes get applied.

* Add more _validate_theme tests

Also fixed some existing ones that no longer tested what we needed them to
when we changed the function to warn instead of raise.

* Do snapshots in both light + dark? (#2900)

* Do snapshots in both light + dark?

* Add snapshots

* Move show_config to the new config_util.py (#2899)

* Move show_config to the new config_util.py

config.py is the longest .py file we have, and I was about to add more
stuff to it, so I thought it was about time to try to split it up.

It's actually pretty hard to move things out of this file given all its
dependencies, but at the very least one of the longest functions was
easily movable, and the new stuff I'm about to add can go into
config_util.

* Just grab lock instead of having a nearly paragraph-long comment

* Design iteration on tooltips (#2915)

* save work

* change tooltipicon

* fix js test

* fix bugs

* fix lint

* Fix issues due to merge

* fix type error

* Warn if the user edits a [server] config option (#2901)

* Drop the ConfigOptions type alias

I've been thinking this alias wasn't really useful given how close it
is to ConfigOption (singular), then I made a mistake because I misread
it for the other :(

* Warn if the user edits a [server] config option

We can't always automatically apply these without a full server restart,
so we should bug the user to reboot streamlit.

Note that warning for all server config options is overly broad, but
it's certainly better to over-warn users than under-warn them in this
case.

* Use parameterized in tests

* Removed unused import

* Don't overwrite config options set via flag on config.toml reload (#2898)

* Don't overwrite config options set via flag on config.toml reload

This was a known issue when we released Theming v0.4, but we decided
that it was minor enough that fixing it could be pushed back until after
that beta was released (presumably it should be rare for a user to set
config options both via flag as well as config.toml file).

* Add clarifying comment

* Rework tests to work with both python 3.6 and 3.8

* Make names in dict comprehension more consistent

* Remove the theme.name option (#2912)

* Remove the theme.name option

This is a task from the product launch review. Instead of letting
developers name their themes, we'll just give any custom theme the name
"Custom Theme" in the theme selector menu.

* Hard code the string "Custom Theme" less

Also rename AUTO_THEME -> AUTO_THEME_NAME

* Wrap up design pass on theming (#2921)

* style scrollbars

* Fix header gradient color

* Style menu and dialogs

* Style everything else!

* Fix small text in dialog

* Make button border use fadedText10, like other borders.

* Fix slider active color.

* Improve "made with streamlit" text

* Remove remaining usage of secondaryColor

* Update enzyme snapshots for SettingsDialog and ThemeCreator

* Get rid of StyledThemeColorPicker

We can just use the ColorPicker directly now that we don't have any styles to
override for it.

* Fix (non-snapshot) js unit tests

* Don't use primary color for background when no radio options

* Don't use primary color for border when checkbox disabled

* Update a TON of snapshots

* Use fadedText colors more where equivalent

* Move derived color computation into a helper and other misc cleanup

* Use shorthand style props for borders where possible

* Remove unnoticed reference to theme.name

This one managed to sneak by since it has no effect on anything in the
case where theme.name is never set.

* Address PR comments

* Get rid of an unnecessary import

Co-authored-by: Thiago Teixeira <thiago@streamlit.io>

* Make tooltips and theming play nicely together (#2939)

The changes in this commit aren't used yet, but they will be very soon
as tooltips are needed for the theme creator dialog.

Aside from pulling in some code that lives in the tooltips feature
branch a bit early, in this commit we also tweak tooltip colors so that
they adjust along with the active theme.

* Stop warn logs from being spammed every time a checkbox is rendered (#2938)

This slipped by when the design pass PR was merged as I guess I never
had a browser console open at the right time :cry:

I happened to notice a ton of warn logs being spammed to the browser
console whenever a checkbox was rendered. Unfortunately, this seems to
be due to the incompatibility of the short/long-hand versions of these
names and the fact that baseweb uses the long versions for some
components, so we don't have much of a choice other than to be very,
very verbose as well.

* Theming markdown and json fixes (#2946)

* Pick a reasonable ReactJson theme depending on background color

* Brighten up markdown links a bit on dark backgrounds

* Add a bunch of st_markdown snapshots

* Don't reload theme on every app rerun (#2937)

* Don't reload theme on every app rerun

Currently, the theme reloads on app rerun when a user changes the value of a
widget, which is because we're applying the theme in the config file
without taking into account whether that theme has actually changed.

This doesn't work well if a user was playing around with colors in the
theme creator dialog, so we should instead only make custom-theme
related changes if a theme config option was actually modified.

* Use join instead of reduce and add a clarifying comment

* Rework Theme Creator UI (#2940)

* Rework Theme Creator UI

* Moves the help text into tooltips
* Has the theme creator be its own dialog instead of being an expandable
  part of the Settings dialog
* Changes the layout of the Theme Creator to have two columns

* Style all dialogs.

* Update snapshots and fix other tests

* Get rid of some now-empty styled components

* Address PR feedback

Co-authored-by: Thiago Teixeira <thiago@streamlit.io>

* Tooltips + theming design changes (#2956)

* save work

* fix padding

* use linkText

* just remove link style

Co-authored-by: karrie <karrie@streamlit.io>
Co-authored-by: Vincent Donato <vincent@streamlit.io>
Co-authored-by: Tim Conkling <tconkling@gmail.com>
Co-authored-by: Thiago Teixeira <thiago@streamlit.io>

schaumb commented:
Yes, there is a difference between the two solution, especially in the main thread, and with any browser connection.
My original solution only works when no connection. The linked solution works always.

Testing the performance: 
There is a `psutil` python package. With that the test can measure CPU times for each thread. 
Possible test case:
- connect a browser on a simple empty streamlit app (where is no live communication, only ws-ping's)
- get cpu time for each thread. `(a)`
- sleep (eg. 1 sec)
- get cpu time for each thread. `(b)`
- assert on `(b) - (a) < threshold`

mesmith027 committed:
Merge pull request #2957 from randyzwitch/favicon_resize

Make favicon smaller

tconkling committed:
FileUploader auto-garbage-collection (#2922)

Overhauls FileUploader deletion, for st.form purposes

- There's no longer a DELETE endpoint for FileUploader, because the semantics don't make sense for FileUploaders that are inside an `st.form`. (We don't want to perform a deletion until the FileUploader's form is submitted.)
- Instead, when a file is uploaded, the server assigns it a new ID, and returns that ID to the client.
- When the client updates its WidgetStateManager value, it passes two things back to the server: the most recent `serverFileID` that the client has seen, and the list of fileIDs of all active uploaded files.
- Each `serverFileID` is a positive integer. IDs are assigned sequentially, so we can compare two files' relative ages: a file with a higher ID is newer than a file with a lower one.
- To delete a file:
  - the client simply removes that file ID from its list of active uploaded files.
  - Whenever file_uploader is updated on the server, there's a new `UploadedFileManager.remove_orphaned_files` function that gets called. This removes files that are unreferenced by the uploader. However, it will *retain* any files that were uploaded more recently than the "mostRecentServerFileId" value - so any uncommitted uploads from a FileUploader in a form will not be prematurely removed.
- `FileUploadClient.uploadFiles` is now `uploadFile`. It just takes a single file at a time, which is how it was being used anyway.

vdonato committed:
Downgrade decamelize to version 4.0.0 (#2967)

Apparently version 5.0.0 breaks in Safari (and the author has no plans of
fixing this as it breaks other features), but the humanize-string
package that relies on it only needs a version more recent than 2.0.0,
so we can downgrade this here.

Ideally, we may want to just get rid of the need for humanize-string
entirely in the near-ish future since its usage is pretty minimal.

vdonato committed:
Downgrade decamelize to version 4.0.0 (#2967)

Apparently version 5.0.0 breaks in Safari (and the author has no plans of
fixing this as it breaks other features), but the humanize-string
package that relies on it only needs a version more recent than 2.0.0,
so we can downgrade this here.

Ideally, we may want to just get rid of the need for humanize-string
entirely in the near-ish future since its usage is pretty minimal.

vdonato committed:
Add theming documentation (#2965)

* Add theming documentation

We add docs for theming to a few places

* A new section in the main concepts doc giving a high-level overview of
  how theming works (for both streamlit app viewers as well as developers).
* An entirely new page describing each of the options available when
  defining a custom theme and providing screenshots with examples for
  the primaryColor and secondaryBackgroundColor options (as those are
  the non-obvious options)
* A new section in the custom components API documentation describing how
  to make your component play nicely with the active theme.

* Tweak capitalization to be consistent with other docs

Minor thing, but other parts of the documentation don't refer to other pages in
the docs as titles / instead write them in all lowercase letters. This commit
changes a link to match that.

vdonato committed:
Add theming documentation (#2965)

* Add theming documentation

We add docs for theming to a few places

* A new section in the main concepts doc giving a high-level overview of
  how theming works (for both streamlit app viewers as well as developers).
* An entirely new page describing each of the options available when
  defining a custom theme and providing screenshots with examples for
  the primaryColor and secondaryBackgroundColor options (as those are
  the non-obvious options)
* A new section in the custom components API documentation describing how
  to make your component play nicely with the active theme.

* Tweak capitalization to be consistent with other docs

Minor thing, but other parts of the documentation don't refer to other pages in
the docs as titles / instead write them in all lowercase letters. This commit
changes a link to match that.

tconkling committed:
Force file_watcher_type to "poll" for secrets.toml (#2968)

Streamlit Sharing makes an app's 'secrets.toml' file available from a special file system that is incompatible with watchdog. So by default, in Sharing, changes to secrets.toml won't be detected at runtime.

With this change, we force the use of PollingFileWatcher (for secrets.toml only) instead of using whatever file_watcher method is set in config.

kantuni committed:
Upgrade xmldom to 0.5.0 or later (#2961)

vdonato committed:
Make some small tweaks to theming docs (#2971)

These are just minor edits for added clarity:

* Added a gif to demonstrate how the Theme Editor works
* Reworded the beginning of theme_options.md to reference the Theming
  section of main_concepts.md. This way the document acts more as a
  reference with the overview of how theming works in main_concepts.
* Mentioned that color names / hex / rgb / hsl (that is, all valid CSS
  colors) are all valid theme color options.

vdonato committed:
Make some small tweaks to theming docs (#2971)

These are just minor edits for added clarity:

* Added a gif to demonstrate how the Theme Editor works
* Reworded the beginning of theme_options.md to reference the Theming
  section of main_concepts.md. This way the document acts more as a
  reference with the overview of how theming works in main_concepts.
* Mentioned that color names / hex / rgb / hsl (that is, all valid CSS
  colors) are all valid theme color options.

vdonato committed:
Only use localStorage when it is available (#2974)

* Only use localStorage when it is available

We ran into an issue with theming on Streamlit Sharing on Chrome
Incognito mode because streamlit apps are run in an iframe so are
considered third party code.

Chrome Incognito disallows third party cookies/data by default, which
was causing the app to fail to load in this case. A (temporary) solution
is to simply avoid trying using `localStorage` when it's unavailable to
us.

* Remove leftover .only

* Grab localStorage in try block too

* Include pathname in key for cached theme

This is necessary as a browser's localStorage is scoped to a domain,
so all apps on Streamlit sharing will have the same key for their cached
theme (so will overwrite each other) if we don't do this.

vdonato committed:
Only use localStorage when it is available (#2974)

* Only use localStorage when it is available

We ran into an issue with theming on Streamlit Sharing on Chrome
Incognito mode because streamlit apps are run in an iframe so are
considered third party code.

Chrome Incognito disallows third party cookies/data by default, which
was causing the app to fail to load in this case. A (temporary) solution
is to simply avoid trying using `localStorage` when it's unavailable to
us.

* Remove leftover .only

* Grab localStorage in try block too

* Include pathname in key for cached theme

This is necessary as a browser's localStorage is scoped to a domain,
so all apps on Streamlit sharing will have the same key for their cached
theme (so will overwrite each other) if we don't do this.

AnOctopus commented:
@schaumb Could you update this PR to be for the linked commit? It sounds like we're going to want that one.

Also, I just tested the linked commit manually (checking htop), and the baseline CPU use for the streamlit server dropped to zero from 3.3% on my laptop, so it appears to work as expected.

kmcgrady committed:
Update change log

kmcgrady committed:
Up version to 0.79.0

schaumb commented:
We need to mock `server` on these failing testcases

Commit:
[tests] mock current server, and assert on enqueue trigger

Commit:
[tests] mock current server, and assert on enqueue trigger

Commit:
[tests] move server mock to patch

Commit:
[tests] move server mock to patch

schaumb commented:
server mock done. We need to bypass mypy problem, because of a tornado bug, which I recently reported:
https://github.com/tornadoweb/tornado/issues/3001

Commit:
Make mypy happy

Commit:
Make mypy happy

Commit:
Merge pull request #2976 from streamlit/release

Release 0.79.0

Commit:
Update submodule commit id

Commit:
Update snapshots for cypress

Commit:
Remove snapshot

Commit:
Remove other snapshot

Commit:
Re-Add snapshots

Commit:
Merge pull request #2984 from kmcgrady/template-repo-update

Update submodule commit id

tconkling committed:
Remove superfluous Pylint "noqa" comments (#3004)

We have a bunch of `# noqa: XYZ`-style comments, which disable pylint warnings.

But we've long since switched from using `pylint` to a combination of `black` (for formatting) and `mypy` (for type correctness), so these no longer mean anything.

vdonato committed:
Stop warn logs from being spammed in dev mode due to override names (#2994)

I noticed a bunch of warn logs in my browser console every time a tooltip is
rendered, and it turns out the reason for this was essentially the same
reason from when this appeared last / was solved by #2938.

This doesn't affect non-dev setups, so I'm not sure how big of a deal this
actually is and whether we should make it a convention to always use the
long names in BaseWeb overrides.

kantuni committed:
[Security] Upgrade is-svg to 4.2.2 or later (#3009)

kantuni committed:
[Security] Upgrade ssri to 8.0.1 or later (#3010)

vdonato committed:
Add snippets about CSS vars and minimum streamlit-component-lib versions (#3008)

The referenced GitHub issue points out that we don't have any examples
for how to actually use the CSS vars that we provide to custom
components. We actually have a snippet about this in our theming launch
blog post, so this commit just adds that bit here verbatim.

Additionally, this adds a note about the minimum supported
streamlit-component-lib version for theming compatibility. This feels a
bit weird to have to do since our docs are versioned, but I think it's
needed in this case since the streamlit-component-lib package version is
decoupled from the PyPI package version that the docs are tied to.

Fixes #2996

vdonato committed:
Add tests to improve coverage around config options set via flag (#2991)

@kantuni pointed out in a previous PR that our test coverage around config
options set via CLI flag was lacking, and now that 0.79/theming has
landed I had a bit of time to improve this.

nrontsis committed:
Improve module reloading (#3001)

- Fixes #3000
- Refactors `update_watched_modules` to reduce complexity

vdonato committed:
Have theme hashes differentiate between being unset and the hash for null (#2993)

* Use better names in processThemeInput and add a comment

* Have theme hashes differentiate between being unset and the hash for null

* Address PR comments

* Clean up grammar in test names

Daniel-DI committed:
Update caching.md (#2999)

fix typo

randyzwitch committed:
Switch theme heading to sentence case

mesmith027 committed:
Merge pull request #3023 from randyzwitch/fix_theme_capitalization

Switch theme heading to sentence case

vdonato committed:
Allow users to click and drag in the color picker (#2992)

It turns out that doing this was actually quite easy as we only need to
use `react-color`s `onChange` instead of `onChangeComplete`. Note that the
corresponding event fires a lot when clicking and dragging, but the
`onChange` handler passed to the ColorPicker component is only called
when the picker popover is closed.

jroes committed:
call _repr_html_ when available

jroes committed:
formatting

tconkling committed:
ComponentInstance: handle iframe changes (#3015)

If a ComponentInstance is reordered in the DOM - which can happen as the result of a script re-run - its iframe will be re-mounted, and get a new `contentWindow`. A Component sends messages to Streamlit via its `contentWindow` -- and since we currently don't detect a change to `contentWindow`, a remounted Component iframe will send messages that just get dropped on the floor by `ComponentRegistry`.

This PR detects `contentWindow` changes during component update and re-registers the component's message listener.

(Most of the hard work of discovery, making a repro-able test case, and finding the solution is thanks to @erumoico: https://github.com/erumoico/streamlit-issue_2720)

Fixes #2720

AnOctopus committed:
Informative repr methods for our python classes (#3027)

* Add repr util method for our classes to use

* Add new generic repr method to almost all classes (excepting those that have reference cycles, which causes infinite recursion)

Commit:
Add fuzzy search to selectbox (#2933)

* add fuzzy search

* clean up code

* fix reverse order

* add unit test

* fix typo

* add beginnings of e2e test

* add newline

* clean things up

* add e2e test

* fix errors & tests

* clean up code

* clean up code some more

AnOctopus commented on lib/streamlit/server/server.py:
```diff
@@ -400,9 +402,13 @@ def _loop_coroutine(self, on_started=None):
 
                 if self._state == State.WAITING_FOR_FIRST_BROWSER:
-                    pass
+                    yield tornado.gen.WaitIterator(
```
Could you add a brief comment explaining what this is doing?

schaumb replied:
I think this code is self documented.
```
if state is Waiting for first browser then
  yield for next of waitIterator (wait must stop , wait has connection)
  ```
  
What kind of comment do you want to see here @AnOctopus ?

AnOctopus replied:
I, having only briefly skimmed the tornado docs, didn't know the semantic meaning of `tornado.gen.maybe_future(self._must_stop.wait()),`, nor that `next()` is the first of those two futures to complete, rather than some sort of toggling between them, for example.
A comment along the lines of "Waits for one of the _must_stop and _has_connection events to occur" would be sufficient. Just enough context to understand the basic control flow, for me and my teammates who aren't familiar with tornado to not have to look it up every time.

schaumb replied:
`maybe_future` function calls needed because of https://github.com/tornadoweb/tornado/issues/3001, it is a workaround, it does not do anything.

So as I understand the `WaitIterator` is not a clear class name, It doesn't explain its own purpose.

I can suggest that It can be used native `asyncio.wait` function, where need to add explicitly  the `FIRST_COMPLETED` constant, and does not need to wrap coroutines/futures (and probably needs no explanation in comment). 

basically:
```
yield tornado.gen.convert_yielded(asyncio.wait([
    self._must_stop.wait(), 
    self._has_connection.wait()
], return_when=asyncio.FIRST_COMPLETED))
```

This modification makes the code obvious?

AnOctopus replied:
Yes, I think that would be sufficient

AnOctopus reviewed (changes requested):
Looks basically good. The only part that isn't pretty obvious what the semantics are is the WaitIterator, so could you add a comment that says what it means?

sachitv committed:
Fix for query param issue with base url (#2894)

This fixes #2887.
Also adds unit tests to App.test.tsx.

vdonato committed:
Have MetricsManager enqueue events when disconnected (#3011)

Currently, MetricsManager only enqueues events when we attempt to send
them before it has been initialized, so trying to send an event when
MetricsManager was initialized but the app is disconnected from the
server results in an error being thrown.

This generally doesn't happen very much in practice (you see it from
time to time when developing streamlit itself when the react dev server
reloads files, but I'm not aware of any other instances where we run into
this).

Some of the theme metrics that I want to add, however, will run into
this when trying to record that a user changed their app's theme while
the app is disconnected, so this needs to be fixed before those metrics
can be added.

Closes #2958

vdonato committed:
Add events to keep track of theme changes and other theme stats (#3012)

Note: This builds off #3011, so that PR should be reviewed first.

With a few small issues with MetricsManager cleaned up, we can now
collect all the theme-related data points we wanted to.

Commit:
Speed up CircleCI more (#3025)

* see if this does anything

* increase parallel instances

* save work

* see if this works

* try turning parallelism back on

* save work

* save work

* save

* remove snapshots

* add snapshots back in

* make some changes

* are you telling me it works if i do this

* save

* fix

* add snapshots in

* run integration tests in parallel

* i think using maxWorkers might be better, try it out

* save

kantuni committed:
Set baseUrl to .

During TypeScript upgrade, `baseUrl: "src"` option introduced some issues. The decision was made to revert back to `baseUrl: "."`.

Commit:
Fix gap not working on Safari (#3042)

* replace gap with a polyfill

* remove function

kantuni committed:
[Security] Upgrade y18n to 4.0.1 or later (#3041)

tconkling committed:
Move secrets out of beta

tconkling committed:
beta_util_test

tconkling committed:
object_beta_warning: handle the subscript operator

tconkling committed:
object_beta_warning: handle every __ magic method

tconkling committed:
Show the beta_warning only a single time per object

tconkling committed:
document getattr

Commit:
Merge pull request #3045 from tconkling/tim/SecretsRelease

streamlit.secrets is no longer beta

vdonato committed:
Remove incorrect markdown table styling (#3038)

* Remove incorrect markdown table styling

We ended up letting a markdown table styling bug slip through where tables are
always displayed with a white background color due to an override that I
guess was left over from the pre-theming days.

Closes #3020

* Add snapshots

jrieke committed:
Fix bokeh example in docs (`legend` arg) (#2907)

tvst committed:
Move docs.streamlit.io to Segment (#3005)

jrieke committed:
Extend docstring of st.image (#3055)

vdonato committed:
Downgrade st.warning in local_sources_watcher to a log (#3050)

I ran into some trouble trying to call st.warning outside of code run by
the ScriptRunner, which made me remember that some code changed in a
recent PR (#3001) did just that.

I double-checked that this code indeed doesn't work as calls that
end up sending Delta protos rely on a ReportContext, which isn't
available in the LocalSourcesWatcher.

Since we can't present the warning in Streamlit itself (which would be
ideal), I just downgraded this to be a warning log for now.

vdonato committed:
Set genericColors properly and make theme defs more consistent (#3051)

We were previously incorrectly reexporting baseTheme.genericColors from
lightTheme instead of overwriting the colors that get changed in
lightTheme. Thankfully, this doesn't make a difference in the light
theme itself but does cause subtle differences in custom themes.

I also tweaked the light/dark theme definitions a bit so they're more
consistent.

jroes committed:
Merge pull request #3029 from streamlit/fix-1117

Call _repr_html_ when available

SimonBiggs committed:
Small fix for `make pylint` command (#3062)

* make 'pylint' actually raise a non-zero exit

* run black so that 'make pylint' passes once more

F1nnM committed:
Added "allow-downloads" to the sandbox attributes (#3053)

* Added "allow-downloads" to the sandbox attributes

This enables components to provide a download button for example for data or images.

* Removed spaces

* Fixed Typo

randyzwitch committed:
Fix pylint errors

randyzwitch committed:
Specify pylint version

randyzwitch committed:
Update sphinx

randyzwitch committed:
See if pylint was red herring

randyzwitch committed:
Is this a red herring also?

mesmith027 committed:
Merge pull request #3076 from randyzwitch/pylint_docs_errors

Fix sphinx version error

Commit:
Fix image galleries (#3044)

* save work

* fix issue

* remove snapshot

* add snapshot back in

* save

* add snapshots back in

Commit:
Merge anchor headers feature branch into develop (#2983)

* Add anchors to Markdown headers (#2513)

* save progress

* save work

* remove junk code

* fix lint errors

* clean up code a bit

* fix incorrect type

* remove debug code

* fix heading numbers

* make requested changes and clean up HeadingWithAnchor

* only scroll once

* fix issues

* fix jslint

* clean things up

* handle edge case

* missing a void

* fix failing test

* add e2e tests

* add e2e tests

* clean up code by using _.once()

* fix lint

* fix dynamic scrolling

* fix typo

* make requested changes

* run formatting and remove unused import

* remove another unused import

* make requested changes

* add nice link icon to left of headers

* make requested design changes (#2726)

* Fix anchor styling to work with dark mode (#3035)

* fix anchor styling

* Revert "fix anchor styling"

This reverts commit cb5d7fdb415e1ac2b1012263c972fe87e0a51bfe.

* fix failing cypress tests

* fix styling

* Add S4A communication for anchor headers (#2982)

* add communication with s4a

* fix bug

* save work

* save work

* remove console

* add tests

* remove useless import

* fix failing tests

* Remove Markdown anchors from sidebar (#3059)

* remove markdown anchors from sidebar

* fix test

* add tests

* Fix failing Cypress test for Markdown Anchors (#3077)

* see if this fixes things

* fix things

* fix

* fix scrolling

* add spacer

* save work

* remove unneeded changes

Commit:
Merge anchor headers feature branch into develop (#2983)

* Add anchors to Markdown headers (#2513)

* save progress

* save work

* remove junk code

* fix lint errors

* clean up code a bit

* fix incorrect type

* remove debug code

* fix heading numbers

* make requested changes and clean up HeadingWithAnchor

* only scroll once

* fix issues

* fix jslint

* clean things up

* handle edge case

* missing a void

* fix failing test

* add e2e tests

* add e2e tests

* clean up code by using _.once()

* fix lint

* fix dynamic scrolling

* fix typo

* make requested changes

* run formatting and remove unused import

* remove another unused import

* make requested changes

* add nice link icon to left of headers

* make requested design changes (#2726)

* Fix anchor styling to work with dark mode (#3035)

* fix anchor styling

* Revert "fix anchor styling"

This reverts commit cb5d7fdb415e1ac2b1012263c972fe87e0a51bfe.

* fix failing cypress tests

* fix styling

* Add S4A communication for anchor headers (#2982)

* add communication with s4a

* fix bug

* save work

* save work

* remove console

* add tests

* remove useless import

* fix failing tests

* Remove Markdown anchors from sidebar (#3059)

* remove markdown anchors from sidebar

* fix test

* add tests

* Fix failing Cypress test for Markdown Anchors (#3077)

* see if this fixes things

* fix things

* fix

* fix scrolling

* add spacer

* save work

* remove unneeded changes

tconkling committed:
Makefile cleanup (#3083)

- Remove `clean-package`, which is misnamed (it doesn't just clean the package files; it cleans _and then_ packages) and unnecessary. Instead, just always clean package files from the `distribution` rule (which is called by `package`).
- Fix `package`'s info string to clarify that it calls `distribution`.
- Fix a number of help strings (grammar & multiline issues)

tconkling committed:
Remove flake8 linting tool (#3085)

We're using black and mypy, not flake8.

Commit:
remove extra star in docs (#3086)

kmcgrady committed:
Revert "Merge pull request #3029 from streamlit/fix-1117"

This reverts commit 96f17464a13969ecdfe31043ab8e875718eb0d10, reversing
changes made to 6684e51bdcf86b2963609f8b0140480ef639b61f.

kmcgrady committed:
Update change log

kmcgrady committed:
Up version to 0.80.0

tconkling committed:
Upgrade bokehjs to 2.2.2, and run `npx browserslist@latest --update-db` (#3090)

- bokehjs: 2.0.0 -> 2.2.2
- ran `npx browserslist@latest --update-db`

This is not a solution for https://github.com/streamlit/streamlit/issues/2156. The latest bokehjs is 2.3.1, but using bokehjs 2.3.0 or 2.3.1 gives a runtime error when the chart is rendered: 

"ReferenceError: Cannot access ‘ColumnarDataSource’ before initialization".

It looks like this is a cyclic dependency issue, and that it's been addressed by bokehjs here: https://github.com/bokeh/bokeh/pull/11140. Presumably, when bokehjs 2.4 releases, we'll be able to it.

(More context in the bokehjs forums, here: https://discourse.bokeh.org/t/problem-importing-bokeh-bokehjs-module-in-javascript-app/7510/12)

kmcgrady committed:
Update Changelog to latest updates

CFrez committed:
Update Num_Input to correct for Type Errors (#3074)

* update to logic order for exception and default

* lint

* correct for default to be float

* add check for no arguments

* shift all_int to  int_args

* add

* add back test of all_ints after assignment of value

* reset to add back accidently deleted test

* switch back proto to all_ints test

* add tests for provided step and default value

* change step test to not match default step value

* moved types to test into tuple

* fix test label

* added max_value test

* added snapshots for new tests

* Replace new snapshots with some generated by CI

Co-authored-by: Vincent Donato <vincent@streamlit.io>

kmcgrady committed:
Modify docs for grammatical error

brendalf committed:
Add import urllib to streamlit hello (#2969) (#3106)

schaumb commented:
ping @AnOctopus

randyzwitch committed:
Put back Sphinx to 3.0.3

randyzwitch committed:
Docutils 0.16

randyzwitch committed:
Add fail_on_warning to build options

AnOctopus committed:
Add s4a message to close active modals (#2893)

* Add s4a message to close active modals

* Handle modal closure information in component update hook

Adding a check that closes the open modal to the componentDidUpdate function ensures it will activate in response to being set by the HOC, and since this resets the command state in the HOC, it avoids the risk of [not rendering the modal] failing to register that the modal had been closed.

AnOctopus commented:
Hey @schaumb. I left a review comment. Just needs a little comment explaining what the WaitIterator is/why we are using it. I could try writing it myself, but since you wrote it I expect you have a better understanding of tornado's api and the change.

mesmith027 committed:
Merge pull request #3126 from randyzwitch/docs_test

Pin docutils to 0.16

mesmith027 committed:
Merge pull request #3126 from randyzwitch/docs_test

Pin docutils to 0.16

kmcgrady committed:
Remove fail on warning

Commit:
Merge pull request #3101 from streamlit/release

Release 0.80.0

schaumb commented:
@AnOctopus  I wrote an answer for that.

![image](https://user-images.githubusercontent.com/6457941/114658331-e8b29080-9cf1-11eb-9585-e680178c6f40.png)

Commit:
Unset fail_on_warning (#3134)

AnOctopus commented:
Ah, it looks like your response was still pending, so I couldn't see it.

Commit:
Fix Markdown Anchors bugs (#3127)

* fix bugs

* remove useless code

* fix styling

* make requested changes

* add trailing newline

* add snapshots back in

* save work

* add snapshots back in

tconkling committed:
Fix regex warnings (#3147)

Fixes 3 regex-related warnings that were thrown when running pytest:

- DeprecationWarning: invalid escape sequence \_
- DeprecationWarning: invalid escape sequence \.
- DeprecationWarning: invalid escape sequence \]

The solve is to use raw strings - which don't process `\` as a Python escape character - instead of normal Python strings.

(Both functions already have test coverage, so this shouldn't break anything.)

tconkling committed:
fix "allow_output_mutation" docstring (remove "not") (#3149)

We warn when return values _are_ mutated, not when they're _not_ mutated.

spicysouvlaki committed:
requirements file updates for share.streamlit.io (#3136)

Updates the docs for pipenv, poetry, and conda support that will be released in the next version of s4t cloud and other changes to how requirements files are parsed.

Commit:
review

Commit:
review

Commit:
Update docs/conf.py (#3135)

tconkling committed:
widgets.py organization + cleanup (#3155)

This is a renaming/reorganization change to widgets.py - no functionality is changed (and so there are no new tests included).

- `Widgets` is now `WidgetStateManager`
- `register_widget` and the `NoValue` type have been moved from `utils.py` into `widgets.py`

jrieke committed:
Update links to `hash_funcs` docs (#3153)

kmcgrady commented:
Thank you @schaumb for the latest changes. I think all that is left is merging in the latest from `develop`. Should fix the merge conflict and fix the docs (there was a bug for a library that broke the docs that was fixed).

SimonBiggs committed:
Adjust media manager to utilise mimetype lib (#3061)

Co-authored-by: Ken McGrady <ken@streamlit.io>

Commit:
remove import

Commit:
Merge branch 'remove_async_busyloop' of https://github.com/FloWide/streamlit into remove_async_busyloop

schaumb commented:
I messed up the merging, I will open a new PR
